### DISCUSSION TITLE
TARDIS biome-setting code

### DISCUSSION THREAD

**Discussion by DalekCraft2** [eccentricdevotion/TARDIS#512] [Q&A] [answered]
I am currently changing much of the code of the plugin to be more up-to-date and simply cleaner. Because the method what is used to set biomes of some TARDIS exteriors includes a deprecated `setBiome(int, int, Biome)` method, I want to add a Y parameter to update it. However, I am confused by how it works; the `for` loops have ranges of -3 <= (X and Z) < 4, which suggests that the biome areas are larger than the 3x3 areas what are necessary for TARDISes. I also do not know how exactly I should implement the Y parameter, since I do not know how exactly I would need to adjust these ranges. May I have an explanation as to why they are set at these values?

**Comment by DalekCraft2** [comment]
![image](https://user-images.githubusercontent.com/50336832/119246314-9c5b3b80-bb4e-11eb-8555-75c056a59d90.png)
Could this work?

**Comment by eccentricdevotion** [comment, marked as answer]
I've tried to update the biome code fairly recently as the method has been deprecated for a while, but never go it to work which is why I left it as is. The issue seems to be in the Minecraft code rather than Bukkit's. Mojang has changed the way that biomes are set in general and even setting the biome for a single block column results in other blocks being changed.

The area being set currently changes depending on which way the TARDIS exterior is facing - it will always be a (x,y,z) 3x4x4 or 4x4x3 cubic area due to the column for the sign:
```
  X       X X X     
X X X     X X X       X X X     X X X     
X X X     X X X     X X X X     X X X X
X X X       X         X X X     X X X     
```

**Reply by eccentricdevotion** [reply]
Also just setting the columns where the TARDIS actually is doesn't work, you sometimes get a TARDIS (police box) with half DEEP_OCEAN biome and half whatever the original biome was...

**Reply by DalekCraft2** [reply]
Well, that sucks. Does that mean that those ranges are just the minimum values what got it to work?

**Reply by eccentricdevotion** [reply]
From my testing, yes.

**Reply by eccentricdevotion** [reply]
Probably the best thing to do it to remove the chameleon presets that rely on changing the biome and just use the ItemFrame based police boxes instead. That way we can just remove the biome setting code for TARDIS exteriors.

**Reply by eccentricdevotion** [reply]
Those textures are really the only ones that rely on Optifine anyway, apart from the Artron furnace (which is pretty obsolete now that there is a blast furnace) and the sonic generator.

**Reply by DalekCraft2** [reply]
How does the Blast Furnace make the Artron Furnace obsolete?

**Reply by eccentricdevotion** [reply]
Both smelt things faster.

**Reply by DalekCraft2** [reply]
Maybe I will just remove it and the CTM exteriors, then.

**Reply by eccentricdevotion** [reply]
Maybe, you'll have to deal with existing exterior preset records in the database though, it won't be as simple as just removing the code / textures...

**Reply by DalekCraft2** [reply]
Luckily, IntelliJ has very nice features for editing many files at once whilst deleting files and renaming variables and such.
